src/core/Ekubo.sol

```diff
@@ -0,0 +1,459 @@
+            }
+            bytes memory encodedPayedAmount =
+                _setOperatorAndCall(msg.sender, data, uint32(IEkuboCallbacks.payCallback.selector), payCallback);
+            assembly ("memory-safe") {
+                // We can skip all the checks performed by `abi.decode` because we know that this is the
+                // verbatim result from `payCallback` and that `payCallback` encoded the payment
+                // correctly.
+                payment := mload(add(0x60, encodedPayedAmount))
+            }
+        }
+    }
+
+    // the mandatory fields are
+    // 2 - sell bps
+    // 1 - pool key tokens case
+    // 32 - config (8 fee, 4 tickSpacing, 20 extension)
+    // 32 - skipAhead
+    uint256 private constant _HOP_DATA_LENGTH = 62;
+
+    uint256 private constant _ADDRESS_MASK = 0x00ffffffffffffffffffffffffffffffffffffffff;
+
+    function locked(bytes calldata data) private returns (bytes memory) {
+        address recipient;
+        uint256 minBuyAmount;
+        uint256 hashMul;
+        uint256 hashMod;
+        bool feeOnTransfer;
+        address payer;
+        (data, recipient, minBuyAmount, hashMul, hashMod, feeOnTransfer, payer) = Decoder.decodeHeader(data);
+
+        // Set up `state` and `notes`. The other values are ancillary and might be used when we need
+        // to settle global sell token debt at the end of swapping.
+        (
+            bytes calldata newData,
+            StateLib.State memory state,
+            NotesLib.Note[] memory notes,
+            ISignatureTransfer.PermitTransferFrom calldata permit,
+            bool isForwarded,
+            bytes calldata sig
+        ) = Decoder.initialize(data, hashMul, hashMod, payer);
+        if (payer != address(this)) {
+            state.globalSell.amount = _permitToSellAmountCalldata(permit);
+        }
+        if (feeOnTransfer) {
+            state.globalSell.amount =
+                _ekuboPay(state.globalSell.token, payer, state.globalSell.amount, permit, isForwarded, sig);
+        }
+        if (state.globalSell.amount == 0) {
+            revert ZeroSellAmount(state.globalSell.token);
+        }
+        state.globalSellAmount = state.globalSell.amount;
+        data = newData;
+
+        PoolKey memory poolKey;
+
+        while (data.length >= _HOP_DATA_LENGTH) {
+            // TODO: Check if BPS is needed
+            uint16 bps;
+            assembly ("memory-safe") {
+                bps := shr(0xf0, calldataload(data.offset))
+
+                data.offset := add(0x02, data.offset)
+                data.length := sub(data.length, 0x02)
+                // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+            }
+
+            data = Decoder.updateState(state, notes, data);
+            // TODO: check the sign convention
+            int256 amountSpecified = int256((state.sell.amount * bps).unsafeDiv(BASIS)).unsafeNeg();
```

Comment:
I'm not sure if this matters (or if it's handled correctly elsewhere) but we should either put this in an `unchecked` block or we should correctly handle overflow in `uint256 -> int256` conversion

Reply:
I kept this TODO from PancakeInfinity implementation. Seems like it was just removed so probably you already decided it was not relevant. See [here](https://github.com/0xProject/0x-settler/pull/311/commits/4ec119bc2e1da8d9caddf1da960f43326ce8d217)

Reply:
Ahh, I'm not referring to the TODO (in PancakeInfinity, I removed it after double checking). I'm referring to casting from uint256 to int256. The upper half of uint256 values cannot be safely cast to int256 and Solidity does not throw an error when you do it

Reply:
So we will revert on errors when the multiplcation overflows, but not when the typecast does. We should either verify that neither are possible (and put it in an `unchecked` block) or handle both by reverting

Reply:
yep, got the issue, but, do I need to double check again 🤔 ? Seems like it is exactly the same case as in PancakeInfinity, that is why I mentioned it

Reply:
Given that all the amounts come from `int128` with the exception of the initial sell amount, I think we can use `unchecked` here. I think that we ought to check that during the header parsing we ought to check that `state.globalSell.amount` fits in an `int128` and then everything ought to be safe after that. What do you think?

Reply:
Sounds good doing so.